qui/updater/utils.py

```diff
@@ -0,0 +1,323 @@
+                            " add attribute `disable_checkboxes`.")
+        if self.disable_checkboxes:
+            return
+        self.disable_checkboxes = True
+        func(self, *args, **kwargs)
+        self.disable_checkboxes = False
+
+    return wrapper
+
+
+def pass_through_event_window(button):
+    """
+    Clicking on header button should activate the button not the header itself.
+    """
+    if not isinstance(button, Gtk.Button):
+        raise TypeError("%r is not a gtk.Button" % button)
+    event_window = button.get_event_window()
+    event_window.set_pass_through(True)
+
+
+class HeaderCheckbox:
+    NONE = 0
+    SAFE = 1
+    EXTENDED = 2
+    ALL = 3
+    SELECTED = 4
+
+    def __init__(
+            self,
+            header_button,
+            allowed: list,
+            callback_all: Callable,
+            callback_some: Callable,
+            callback_none: Callable
+    ):
+        self.header_button = header_button
+        self.state = HeaderCheckbox.SAFE
+        self._allowed = allowed
+        self.callback_all = callback_all
+        self.callback_some = callback_some
+        self.callback_none = callback_none
+
+    @property
+    def allowed(self):
+        if self.state == HeaderCheckbox.ALL:
+            return self._allowed[:]
+        if self.state == HeaderCheckbox.SAFE:
+            return self._allowed[:1]
+        if self.state == HeaderCheckbox.EXTENDED:
+            return self._allowed[:2]
+        if self.state == HeaderCheckbox.NONE:
+            return ()
+
+    def set_buttons(self, *args):
+        if self.state == HeaderCheckbox.ALL:
+            self.header_button.set_inconsistent(False)
+            self.header_button.set_active(True)
+            self.callback_all(*args)
+        elif self.state == HeaderCheckbox.NONE:
+            self.header_button.set_inconsistent(False)
+            self.header_button.set_active(False)
+            self.callback_none(*args)
+        else:
+            self.header_button.set_inconsistent(True)
+            self.callback_some(*args)
+
+    def next_state(self):
+        self.state = (self.state + 1) % 4  # SELECTED is skipped
+
+
+def on_head_checkbox_toggled(list_store, head_checkbox, select_rows):
+    if len(list_store) == 0:  # to avoid infinite loop
+        head_checkbox.state = HeaderCheckbox.NONE
+        selected_num = 0
+    else:
+        selected_num = selected_num_old = sum(
+            row.selected for row in list_store)
+        while selected_num == selected_num_old:
+            head_checkbox.next_state()
+            select_rows()
+            selected_num = sum(
+                row.selected for row in list_store)
+    plural = "s" if selected_num > 1 else ""
+    head_checkbox.set_buttons(plural, selected_num)
+
+
+class QubeClass(Enum):
+    """
+    Sorting order by vm type.
+    """
+    AdminVM = 0
+    TemplateVM = 1
+    StandaloneVM = 2
+    AppVM = 3
+    DispVM = 4
+
+
+class QubeLabel(Enum):
```

Comment:
qubesadmin.label object has index already, and the sorting there is used in other places like qube manager, I think it should / can be used here, too